File: backend/src/AIYouTubeFactory.Core/Models/ContentModels.cs
namespace AIYouTubeFactory.Core.Models;

public class UploadedDocument
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public string FileName { get; set; } = string.Empty;
    public string FileType { get; set; } = string.Empty; // pdf, pptx, docx, md
    public string ExtractedText { get; set; } = string.Empty;
    public List<string> Sections { get; set; } = new();
    public string Topic { get; set; } = string.Empty;
    public DateTime UploadedAt { get; set; } = DateTime.UtcNow;
}

public class ContentGenerationRequest
{
    public Guid DocumentId { get; set; }
    public string Topic { get; set; } = string.Empty;
    public int YouTubeVideoCount { get; set; } = 10;
    public int ShortsCount { get; set; } = 20;
    public int LinkedInPostCount { get; set; } = 5;
    public int TwitterThreadCount { get; set; } = 5;
    public bool GenerateThumbnailPrompts { get; set; } = true;
    public bool GenerateSEO { get; set; } = true;
    public string TargetAudience { get; set; } = "developers and tech enthusiasts";
    public string ContentStyle { get; set; } = "educational"; // educational, entertaining, inspirational
}

public class ContentGenerationResult
{
    public Guid SessionId { get; set; } = Guid.NewGuid();
    public string Topic { get; set; } = string.Empty;
    public List<YouTubeVideoScript> YouTubeScripts { get; set; } = new();
    public List<ShortsScript> ShortsScripts { get; set; } = new();
    public List<LinkedInPost> LinkedInPosts { get; set; } = new();
    public List<TwitterThread> TwitterThreads { get; set; } = new();
    public List<ThumbnailPrompt> ThumbnailPrompts { get; set; } = new();
    public VideoContentPlan ContentPlan { get; set; } = new();
    public DateTime GeneratedAt { get; set; } = DateTime.UtcNow;
}

public class YouTubeVideoScript
{
    public int Index { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string Hook { get; set; } = string.Empty;
    public string Introduction { get; set; } = string.Empty;
    public List<ScriptSection> MainContent { get; set; } = new();
    public string CallToAction { get; set; } = string.Empty;
    public string Outro { get; set; } = string.Empty;
    public int EstimatedDurationMinutes { get; set; }
    public SEOData SEO { get; set; } = new();
    public ThumbnailPrompt ThumbnailPrompt { get; set; } = new();
}

public class ScriptSection
{
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public int DurationSeconds { get; set; }
    public string VisualNote { get; set; } = string.Empty;
}

public class ShortsScript
{
    public int Index { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Hook { get; set; } = string.Empty;
    public string MainPoint { get; set; } = string.Empty;
    public string CallToAction { get; set; } = string.Empty;
    public int DurationSeconds { get; set; } = 60;
    public List<string> Hashtags { get; set; } = new();
    public string VisualConcept { get; set; } = string.Empty;
}

public class LinkedInPost
{
    public int Index { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public List<string> Hashtags { get; set; } = new();
    public string CallToAction { get; set; } = string.Empty;
    public string PostType { get; set; } = string.Empty; // article, thought-leadership, tips, story
}

public class TwitterThread
{
    public int Index { get; set; }
    public string Topic { get; set; } = string.Empty;
    public List<string> Tweets { get; set; } = new();
    public List<string> Hashtags { get; set; } = new();
}

public class ThumbnailPrompt
{
    public string VideoTitle { get; set; } = string.Empty;
    public string MainText { get; set; } = string.Empty;
    public string BackgroundDescription { get; set; } = string.Empty;
    public string ColorScheme { get; set; } = string.Empty;
    public string FaceExpression { get; set; } = string.Empty;
    public string DallePrompt { get; set; } = string.Empty;
    public string MidjourneyPrompt { get; set; } = string.Empty;
}

public class SEOData
{
    public string PrimaryKeyword { get; set; } = string.Empty;
    public List<string> SecondaryKeywords { get; set; } = new();
    public List<string> Tags { get; set; } = new();
    public string OptimizedTitle { get; set; } = string.Empty;
    public string OptimizedDescription { get; set; } = string.Empty;
    public List<string> Chapters { get; set; } = new();
}

public class VideoContentPlan
{
    public string OverallStrategy { get; set; } = string.Empty;
    public List<string> ContentPillars { get; set; } = new();
    public List<PublishingScheduleItem> PublishingSchedule { get; set; } = new();
    public List<string> SeriesIdeas { get; set; } = new();
    public string ChannelGrowthStrategy { get; set; } = string.Empty;
}

public class PublishingScheduleItem
{
    public int WeekNumber { get; set; }
    public string ContentType { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Platform { get; set; } = string.Empty;
    public string PublishDay { get; set; } = string.Empty;
}

public class AgentProgressUpdate
{
    public string AgentName { get; set; } = string.Empty;
    public string Status { get; set; } = string.Empty; // running, completed, error
    public string Message { get; set; } = string.Empty;
    public int ProgressPercent { get; set; }
    public object? Data { get; set; }
}

File: frontend/src/app/core/models/content.models.ts
export interface UploadResponse {
  documentId: string;
  fileName: string;
  topic: string;
  extractedChars: number;
  sectionCount: number;
  preview: string;
}

export interface ContentGenerationRequest {
  documentId: string;
  topic: string;
  youTubeVideoCount: number;
  shortsCount: number;
  linkedInPostCount: number;
  twitterThreadCount: number;
  generateThumbnailPrompts: boolean;
  generateSEO: boolean;
  targetAudience: string;
  contentStyle: string;
}

export interface ContentGenerationResult {
  sessionId: string;
  topic: string;
  youTubeScripts: YouTubeVideoScript[];
  shortsScripts: ShortsScript[];
  linkedInPosts: LinkedInPost[];
  twitterThreads: TwitterThread[];
  thumbnailPrompts: ThumbnailPrompt[];
  contentPlan: VideoContentPlan;
  generatedAt: string;
}

export interface YouTubeVideoScript {
  index: number;
  title: string;
  description: string;
  hook: string;
  introduction: string;
  mainContent: ScriptSection[];
  callToAction: string;
  outro: string;
  estimatedDurationMinutes: number;
  seo: SEOData;
  thumbnailPrompt: ThumbnailPrompt;
}

export interface ScriptSection {
  title: string;
  content: string;
  durationSeconds: number;
  visualNote: string;
}

export interface ShortsScript {
  index: number;
  title: string;
  hook: string;
  mainPoint: string;
  callToAction: string;
  durationSeconds: number;
  hashtags: string[];
  visualConcept: string;
}

export interface LinkedInPost {
  index: number;
  title: string;
  content: string;
  hashtags: string[];
  callToAction: string;
  postType: string;
}

export interface TwitterThread {
  index: number;
  topic: string;
  tweets: string[];
  hashtags: string[];
}

export interface ThumbnailPrompt {
  videoTitle: string;
  mainText: string;
  backgroundDescription: string;
  colorScheme: string;
  faceExpression: string;
  dallePrompt: string;
  midjourneyPrompt: string;
}

export interface SEOData {
  primaryKeyword: string;
  secondaryKeywords: string[];
  tags: string[];
  optimizedTitle: string;
  optimizedDescription: string;
  chapters: string[];
}

export interface VideoContentPlan {
  overallStrategy: string;
  contentPillars: string[];
  publishingSchedule: PublishingScheduleItem[];
  seriesIdeas: string[];
  channelGrowthStrategy: string;
}

export interface PublishingScheduleItem {
  weekNumber: number;
  contentType: string;
  title: string;
  platform: string;
  publishDay: string;
}

export interface AgentProgressUpdate {
  agentName: string;
  status: 'running' | 'completed' | 'error';
  message: string;
  progressPercent: number;
  data?: any;
}

export interface GenerationSession {
  sessionId: string;
  documentId: string;
  topic: string;
  status: 'pending' | 'running' | 'completed' | 'error';
  agentUpdates: AgentProgressUpdate[];
  result?: ContentGenerationResult;
  startedAt: Date;
}

// ── Video Generation Models ───────────────────────────────────────────────────
export interface VideoGenerationRequest {
  sessionId:               string;
  scriptIndex:             number;
  voiceName:               string;
  voiceStyle:              string;
  speechRate:              number;
  generateSubtitles:       boolean;
  generateImages:          boolean;
  imageStyle:              string;
  outputFormat:            string;
  resolution:              number; // 0=720p 1=1080p 2=4K
}

export interface VideoGenerationResponse {
  videoSessionId: string;
  message:        string;
  scriptTitle:    string;
}

export interface GeneratedVideo {
  id:              string;
  title:           string;
  status:          VideoStatus;
  durationSeconds: number;
  fileSizeBytes:   number;
  resolution:      string;
  sceneCount:      number;
  hasSubtitles:    boolean;
  createdAt:       string;
  errorMessage?:   string;
  downloadUrl?:    string;
  subtitleUrl?:    string;
  thumbnailUrl?:   string;
}

export type VideoStatus =
  | 'Pending'
  | 'GeneratingStoryboard'
  | 'GeneratingImages'
  | 'GeneratingVoice'
  | 'GeneratingSubtitles'
  | 'ComposingVideo'
  | 'Completed'
  | 'Failed';

export interface VideoProgressUpdate {
  stage:           string;
  message:         string;
  progressPercent: number;
  status:          VideoStatus;
  previewImagePath?: string;
}
